frontend/lib/src/components/widgets/DataFrame/DataFrame.tsx

```diff
@@ -263,50 +241,150 @@ function DataFrame({
     useColumnSort(originalNumRows, originalColumns, getOriginalCellContent)
 
   /**
-   * This callback should be called after any edits have been applied to the data.
-   * It will finish up the editing by updating the number of rows, clearing the selection,
-   * and triggering a rerun of the script.
+   * This callback is used to synchronize the selection state with the state
+   * of the widget state of the component. This might also send a rerun message
+   * to the backend if the selection state has changed.
    *
-   * @param clearSelections - Whether to clear the selection. This is usually done after deleting rows.
-   * @param triggerRerun - Whether to trigger a rerun of the script after applying edits
+   * @param newSelection - The new selection state
    */
-  const applyEdits = React.useCallback(
-    (clearSelections = false, triggerRerun = true) => {
-      if (numRows !== editingState.current.getNumRows()) {
-        // Reset the number of rows if it has been changed in the editing state
-        setNumRows(editingState.current.getNumRows())
+  // The debounce method doesn't allow dependency inspection. Therefore, we
+  // need to disable the eslint rule for exhaustive-deps.
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  const syncSelectionState = React.useCallback(
+    // Use debounce to prevent rapid updates to the widget state.
+    debounce(DEBOUNCE_TIME_MS, (newSelection: GridSelection) => {
+      // If we want to support selections also with in the editable mode,
+      // we would need to integrate the `applyEdits` and `syncSelections` functions
+      // into a single function that updates the widget state with both the editing
+      // state and the selection state.
+
+      const selectionState: DataframeState = {
+        select: {
+          rows: [] as number[],
+          columns: [] as string[],
+        },
       }
 
-      if (clearSelections) {
-        clearSelection()
+      selectionState.select.rows = newSelection.rows.toArray().map(row => {
+        return getOriginalIndex(row)
+      })
+      selectionState.select.columns = newSelection.columns
+        .toArray()
+        .map(columnIdx => {
+          return getColumnName(originalColumns[columnIdx])
```

Comment:
We are using it here since we decided to use the same identifying attributes for edits with data editor. And the general decision is to use what is most descriptive and easiest to use in the python backend. For columns that's the name, for rows we don't really have something equivalent. We evaluated to use the row index (index columns),  but there is a lot of complexity around that ... and most dataframes don't even have an index defined.